### PR TITLE
feat: Add `Hellinger` distance

### PR DESCRIPTION
**Pull request by robinholzi**
# Motivation

Support `Hellinger distance` even though it's not directly supported by `evidently` or `alibi-detect`.

### CODE REVIEW

File: modyn/config/schema/pipeline/trigger/drift/evidently.py [hunk 1]
```diff
@@ -5,39 +5,52 @@
...
 
 
-class EvidentlyModelDriftMetric(_EvidentlyBaseDriftMetric):
+class EvidentlyModelDriftMetric(_EvidentlyPcaDriftMetric):
     id: Literal["EvidentlyModelDriftMetric"] = Field("EvidentlyModelDriftMetric")
     bootstrap: bool = Field(False)
-    quantile_probability: float = 0.95
-    threshold: float = Field(0.55)
+    quantile_probability: float = Field(
+        0.95, description="Threshold for the evidently api, unused in modyn. Usage discourged."
```
Comment (MaxiBoether) [inline modyn/config/schema/pipeline/trigger/drift/evidently.py:16]: Copy&paste description from threshold?

File: modyn/config/schema/pipeline/trigger/drift/evidently.py [hunk 2]
```diff
@@ -5,39 +5,52 @@
...
-    component_stattest_threshold: float = Field(0.1)
-    threshold: float = Field(0.2)
+    component_stattest_threshold: float = Field(
+        0.1, description="Threshold for the evidently api, unused in modyn. Usage discourged."
+    )
+    threshold: float = Field(0.2, description="Threshold for the evidently api, unused in modyn. Usage discourged.")
 
 
-class EvidentlySimpleDistanceDriftMetric(_EvidentlyBaseDriftMetric):
+class EvidentlySimpleDistanceDriftMetric(_EvidentlyPcaDriftMetric):
```
Comment (MaxiBoether) [inline modyn/config/schema/pipeline/trigger/drift/evidently.py:33]: I don't get yet why everything is a Pca drift metric from an inheritance sense. I suppose if num_pca_components is none, no PCA is applied?

File: modyn/supervisor/internal/triggers/drift/detector/evidently.py [hunk 3]
```diff
@@ -110,6 +118,7 @@ def _get_evidently_metrics(
             EmbeddingsDriftMetric(EVIDENTLY_COLUMN_MAPPING_NAME, _evidently_metric_factory(config)),
         )
         for metric_ref, config in metrics_config.items()
+        if config.id not in ["EvidentlyHellingerDistanceDriftMetric"]
```
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/drift/detector/evidently.py:121]: can we make this a global in this file like `SPECIAL_METRICS` or something?

File: modyn/supervisor/internal/triggers/drift/detector/evidently.py [hunk 4]
```diff
@@ -141,3 +150,25 @@ def _evidently_metric_factory(config: EvidentlyDriftMetric) -> EmbeddingsDriftMe
...
+    embeddings_ref: pd.DataFrame,
+    embeddings_cur: pd.DataFrame,
+) -> dict[str, MetricResult]:
+    metric_results: dict[str, MetricResult] = {}
+    for metric_ref, config in configs.items():
+        if config.id == "EvidentlyHellingerDistanceDriftMetric":
+            column_distances = [
+                # [0]: Hellinger distance, [1]: decision with dummy threshold (False)
+                _hellinger_distance(embeddings_ref[c], embeddings_cur[c], ColumnType.Numerical, 0)[0]
+                for c in embeddings_ref.columns
```
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/drift/detector/evidently.py:166]: Each column is a data point, and the rows in this column vector are the embedding dimensions? (in my head it was the other way around just want to make sure we supply it the right way)

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thank you! Just had small comments / naming questions, after this we can merge

File: modyn/config/schema/pipeline/trigger/drift/evidently.py (hunk 1 above)
Comment (robinholzi) [reply modyn/config/schema/pipeline/trigger/drift/evidently.py:16]: no the quantile probability is also a threshold, but in the probability space

File: modyn/config/schema/pipeline/trigger/drift/evidently.py (hunk 2 above)
Comment (robinholzi) [reply modyn/config/schema/pipeline/trigger/drift/evidently.py:33]: Renamed to mixin, it's just for avoiding have to duplicate the field everywhere. And yes, none is no pca. We don't use that in our experiments though.

File: modyn/supervisor/internal/triggers/drift/detector/evidently.py (hunk 3 above)
Comment (robinholzi) [reply modyn/supervisor/internal/triggers/drift/detector/evidently.py:121]: sure

**Review by MaxiBoether — APPROVED** [review]

**Pull request merged**